src/test/java/org/cli/CliTests.java

```diff
@@ -0,0 +1,178 @@
+            boolean passed = false;
+            try {
+                host.run();
+            } catch (EarlyExitException e) {
+                passed = true;
+            }
+            assertTrue(passed);
+        }
+
+        private ApplicationHost buildApplicationHost(String applicationPath,
+                                                     String restoreResource,
+                                                     Class<E> clazz,
+                                                     String steps) throws NoSuchMethodException, IllegalAccessException, InvocationTargetException, InstantiationException {
+            ClassLoader classLoader = getClass().getClassLoader();
+            File appFile = new File(classLoader.getResource(applicationPath).getFile());
+            String resourcePath = appFile.getAbsolutePath();
+            PrototypeFactory prototypeFactory = new LivePrototypeFactory();
+
+            CommCareConfigEngine engine = CliCommand.configureApp(resourcePath, prototypeFactory);
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
+            PrintStream outStream = new PrintStream(baos);
+
+            Constructor<E> ctor = clazz.getConstructor(String.class, ByteArrayOutputStream.class);
+            CliTestReader reader = ctor.newInstance(steps, baos);
+
+            ApplicationHost host = new ApplicationHost(engine, prototypeFactory, reader, outStream);
+            File restoreFile = new File(classLoader.getResource(restoreResource).getFile());
+            String restorePath = restoreFile.getAbsolutePath();
+            host.setRestoreToLocalFile(restorePath);
+            return host;
+        }
+    }
+
+    @Test
+    public void testConstraintsForm() throws Exception {
+        // Start a basic form
+        new CliTestRun<>("basic_app/basic.ccz",
+                "case_create_basic.xml",
+                BasicTestReader.class,
+                "1 0 \n");
+    }
+
+    @Test
+    public void testCaseSelection() throws Exception {
+        // Perform case selection
+        new CliTestRun<>("basic_app/basic.ccz",
+                "basic_app/restore.xml",
+                CaseTestReader.class,
+                "2 1 5 \n");
+    }
+
+    static abstract class CliTestReader extends BufferedReader {
+
+        private String[] args;
+        private int index;
+        private ByteArrayOutputStream outStream;
+
+        CliTestReader(String args, ByteArrayOutputStream outStream) {
+            super(new StringReader("Unused dummy reader"));
+            this.args = args.split(" ");
+            this.outStream = outStream;
+        }
+
+        @Override
+        public String readLine() throws IOException {
+            String output = new String(outStream.toByteArray(), StandardCharsets.UTF_8);
+            processLine(index, output);
+            String ret = args[index++];
+            outStream.reset();
+            return ret;
```

Comment:
I'm confused why this method returns 1 of your steps

Reply:
Added a comment explaining this class